### PR TITLE
fix: models get_or_create keyerror

### PR DESCRIPTION
**Pull request by jiangying000**
#1583

### CODE REVIEW

**Comment by abondar** [comment]
Could you please rebase your PR on actual develop branch?

**Comment by jiangying000** [comment]
> Could you please rebase your PR on actual develop branch? Also please add small test on it and info about change to changelog

yes, i did this

File: tortoise/models.py
Comment (waketzheng) [inline tortoise/models.py:1]: 1. `kwargs` is 'Query parameters', `defaults` is 'Default values used to update the object', when they have same key with different values, value of `kwargs` should be covered by `defaults`. That is to say code must be: `merged_defaults = {**kwargs, **defaults}` 
2. merged_defaults = ... should be in front of `try:` (just put it after line 1082)
3. `return ...` shoud be stay as one line, do not make it into three lines.

Comment (jiangying000) [reply tortoise/models.py:1]: thanks, done

Comment (abondar) [reply tortoise/models.py:1]: Even though that is how it works in django - for me it seems strange that `default` value overrides explicit value

I would expect opposite behaviour of this

I dunno if we should proceed with this change, as leaving merging on user side seems like more transparent behaviour

Comment (jiangying000) [reply tortoise/models.py:1]: i would also expect explicit kv overrides the default and do merge rather than throw err is ok for me

Comment (abondar) [reply tortoise/models.py:1]: It seems to me - in your case you should do explicit `get_or_create` without `defaults` and them make usual `update()` for returned object
That shouldn't affect performance, as it is basically what we already do

Django also, from 5.0, introduced separate `create_defaults` param for `update_to_create` for similar cases to make them less confusing
May be we should look into doing it too, but it seems there could be some issues in tortoise to use `select_for_update` with `get_or_create`

Comment (waketzheng) [reply tortoise/models.py:1]: > Even though that is how it works in django - for me it seems strange that `default` value overrides explicit value
> 
> I would expect opposite behaviour of this
> 
> I dunno if we should proceed with this change, as leaving merging on user side seems like more transparent behaviour

Our team use tortoise-orm for a long time, and we are proficient in django. So we prefer it have the same result with django.

Or maybe an exception can be raised when kwargs and defaults have same key with different values:
```py
for key in (defaults.keys() & kwargs.keys()):
    if (default_value := defaults[key]) != (query_value := kwargs[key]):
        raise tortoise.exceptions.ParamError(f'Conflict value with {key=}: {default_value=}  vs {query_value=}')
```

Comment (abondar) [reply tortoise/models.py:1]: > Or maybe an exception can be raised when kwargs and defaults have same key with different values:

I would prefer that, as django behaviour seems confusing to me here, but doing solution opposite to them - would be also confusing

Comment (jiangying000) [reply tortoise/models.py:1]: > > Or maybe an exception can be raised when kwargs and defaults have same key with different values:
> 
> I would prefer that, as django behaviour seems confusing to me here, but doing solution opposite to them - would be also confusing

ok, updated

File: tests/test_model_methods.py
Comment (waketzheng) [inline tests/test_model_methods.py:1]: Should add case about `kwargs` and `defaults` have same key and same value

File: tortoise/models.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
 from pypika.terms import Term
 from typing_extensions import Self
 
+import tortoise
```
Comment (waketzheng) [inline tortoise/models.py:26]: Remove this line

File: tortoise/models.py [hunk 2]
```diff
@@ -1073,17 +1073,23 @@ async def get_or_create(
...
+        except DoesNotExist:
             try:
-                return (
-                    await cls.select_for_update().filter(**kwargs).using_db(connection).get(),
-                    False,
-                )
-            except DoesNotExist:
+                for key in (defaults.keys() & kwargs.keys()):
+                    if (default_value := defaults[key]) != (query_value := kwargs[key]):
+                        raise tortoise.exceptions.ParamsError(
```
Comment (waketzheng) [inline tortoise/models.py:1082]: Change it to:
```py
raise ParamsError("...")
```
Then run `make style` to reformat this file

File: tortoise/models.py
Comment (waketzheng) [inline tortoise/models.py:1]: image

**Comment by waketzheng** [comment]
@jiangying000 How is that?

**Comment by jiangying000** [comment]
> @jiangying000 How is that?

sorry, no progress yet, too busy dealing with my employment job tasks

**Comment by jiangying000** [comment]
feel free to take on this pr, or close it

**Comment by waketzheng** [comment]
> feel free to take on this pr, or close it

I have make a pr, could you merge it into your develop branch?
https://github.com/jiangying000/tortoise-orm/pull/1

**Comment by jiangying000** [comment]
> > feel free to take on this pr, or close it
> 
> I have make a pr, could you merge it into your develop branch? [jiangying000#1](https://github.com/jiangying000/tortoise-orm/pull/1)

done, thanks

**Comment by jiangying000** [comment]
> > feel free to take on this pr, or close it
> 
> I have make a pr, could you merge it into your develop branch? [jiangying000#1](https://github.com/jiangying000/tortoise-orm/pull/1)

done, thanks

**Review by waketzheng — APPROVED** [review]

**Comment by waketzheng** [comment]
@abondar Could you review this?

**Comment by abondar** [comment]
@waketzheng seems like there is some issue 
tests hangs up somewhere on ODBC tests

Also we would need to update changelog to put change in new section

**Comment by waketzheng** [comment]
> @waketzheng seems like there is some issue tests hangs up somewhere on ODBC tests
> 
> Also we would need to update changelog to put change in new section

@abondar done

**Comment by waketzheng** [comment]
@abondar Cloud you review it?

**Review by abondar — APPROVED** [review]

**Pull request merged**